fennel/expr/expr.py

```diff
@@ -0,0 +1,748 @@
+from __future__ import annotations
+from enum import Enum
+from dataclasses import dataclass
+from typing import Any, Callable, Type, Optional
+
+import json
+
+from fennel.dtypes.dtypes import FENNEL_STRUCT
+
+
+class InvalidExprException(Exception):
+    pass
+
+
+class Expr(object):
+    def __init__(self, root=None, dtype=None):
+        self.nodeid = id(self)
+        self.inline = False
+        self.out_edges = []
+        self.root = self if root is None else root
+        self.dtype = dtype
+
+    def edge(self, node):
+        self.out_edges.append(node)
+
+    def astype(self, dtype: Type) -> Expr:
+        self.dtype = dtype
+        return self
```

Comment:
JFYI - the way this is written, someone could put `astype` in the middle of the expression or put `astype` twice. You may want to consider an alternative where there are two classes `Expr` and `TypedExpr` or something like that. TypedExpr has no other methods so is terminal. astype on Expr takes you to TypedExpr.